File: server/app/core/models/handbook.py
from datetime import date, datetime
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


HandbookStatus = Literal["draft", "active", "archived"]
HandbookMode = Literal["single_state", "multi_state"]
HandbookSourceType = Literal["template", "upload"]
HandbookSectionType = Literal["core", "state", "custom", "uploaded"]
HandbookChangeStatus = Literal["pending", "accepted", "rejected"]


class HandbookScopeInput(BaseModel):
    state: str = Field(..., min_length=2, max_length=2)
    city: Optional[str] = None
    zipcode: Optional[str] = None
    location_id: Optional[UUID] = None


class HandbookSectionInput(BaseModel):
    section_key: str = Field(..., min_length=2, max_length=120)
    title: str = Field(..., min_length=2, max_length=255)
    content: str = ""
    section_order: int = 0
    section_type: HandbookSectionType = "custom"
    jurisdiction_scope: Optional[dict] = None


class CompanyHandbookProfileInput(BaseModel):
    legal_name: str = Field(..., min_length=1, max_length=255)
    dba: Optional[str] = Field(default=None, max_length=255)
    ceo_or_president: str = Field(..., min_length=1, max_length=255)
    headcount: Optional[int] = None
    remote_workers: bool = False
    minors: bool = False
    tipped_employees: bool = False
    union_employees: bool = False
    federal_contracts: bool = False
    group_health_insurance: bool = False
    background_checks: bool = False
    hourly_employees: bool = True
    salaried_employees: bool = False
    commissioned_employees: bool = False
    tip_pooling: bool = False


class HandbookCreateRequest(BaseModel):
    title: str = Field(..., min_length=2, max_length=500)
    mode: HandbookMode = "single_state"
    source_type: HandbookSourceType = "template"
    industry: Optional[str] = Field(default=None, max_length=120)
    scopes: list[HandbookScopeInput]
    profile: CompanyHandbookProfileInput
    custom_sections: list[HandbookSectionInput] = Field(default_factory=list)
    file_url: Optional[str] = None
    file_name: Optional[str] = None
    create_from_template: bool = True

    @model_validator(mode="after")
    def validate_shape(self):
        if self.mode == "single_state" and len(self.scopes) != 1:
            raise ValueError("Single-state handbooks must have exactly one scope")
        if self.mode == "multi_state" and len(self.scopes) < 2:
            raise ValueError("Multi-state handbooks must include at least two scopes")
        if self.source_type == "upload" and not self.file_url:
            raise ValueError("Uploaded handbooks require file_url")
        return self


class HandbookUpdateRequest(BaseModel):
    title: Optional[str] = None
    mode: Optional[HandbookMode] = None
    scopes: Optional[list[HandbookScopeInput]] = None
    profile: Optional[CompanyHandbookProfileInput] = None
    sections: Optional[list[HandbookSectionInput]] = None
    file_url: Optional[str] = None
    file_name: Optional[str] = None

    @model_validator(mode="after")
    def validate_mode_scopes(self):
        if self.mode == "single_state" and self.scopes is not None and len(self.scopes) != 1:
            raise ValueError("Single-state handbooks must have exactly one scope")
        if self.mode == "multi_state" and self.scopes is not None and len(self.scopes) < 2:
            raise ValueError("Multi-state handbooks must include at least two scopes")
        return self


class HandbookScopeResponse(BaseModel):
    id: UUID
    state: str
    city: Optional[str] = None
    zipcode: Optional[str] = None
    location_id: Optional[UUID] = None


class HandbookSectionResponse(BaseModel):
    id: UUID
    section_key: str
    title: str
    content: str
    section_order: int
    section_type: HandbookSectionType
    jurisdiction_scope: dict = Field(default_factory=dict)


class CompanyHandbookProfileResponse(CompanyHandbookProfileInput):
    company_id: UUID
    updated_by: Optional[UUID] = None
    updated_at: datetime


class HandbookListItemResponse(BaseModel):
    id: UUID
    title: str
    status: HandbookStatus
    mode: HandbookMode
    source_type: HandbookSourceType
    active_version: int
    scope_states: list[str]
    pending_changes_count: int = 0
    updated_at: datetime
    published_at: Optional[datetime] = None
    created_at: datetime


class HandbookDetailResponse(BaseModel):
    id: UUID
    company_id: UUID
    title: str
    status: HandbookStatus
    mode: HandbookMode
    source_type: HandbookSourceType
    active_version: int
    file_url: Optional[str] = None
    file_name: Optional[str] = None
    scopes: list[HandbookScopeResponse]
    profile: CompanyHandbookProfileResponse
    sections: list[HandbookSectionResponse]
    created_at: datetime
    updated_at: datetime
    published_at: Optional[datetime] = None
    created_by: Optional[UUID] = None


class HandbookChangeRequestResponse(BaseModel):
    id: UUID
    handbook_id: UUID
    handbook_version_id: UUID
    alert_id: Optional[UUID] = None
    section_key: Optional[str] = None
    old_content: Optional[str] = None
    proposed_content: str
    rationale: Optional[str] = None
    source_url: Optional[str] = None
    effective_date: Optional[date] = None
    status: HandbookChangeStatus
    resolved_by: Optional[UUID] = None
    resolved_at: Optional[datetime] = None
    created_at: datetime


class HandbookPublishResponse(BaseModel):
    id: UUID
    status: HandbookStatus
    active_version: int
    published_at: Optional[datetime] = None


class HandbookDistributionResponse(BaseModel):
    handbook_id: UUID
    handbook_version: int
    assigned_count: int
    skipped_existing_count: int
    distributed_at: datetime


class HandbookAcknowledgementSummary(BaseModel):
    handbook_id: UUID
    handbook_version: int
    assigned_count: int
    signed_count: int
    pending_count: int
    expired_count: int


class HandbookGuidedQuestion(BaseModel):
    id: str = Field(..., min_length=2, max_length=80)
    question: str = Field(..., min_length=4, max_length=500)
    placeholder: Optional[str] = Field(default=None, max_length=255)
    required: bool = True


class HandbookGuidedSectionSuggestion(BaseModel):
    section_key: str = Field(..., min_length=2, max_length=120)
    title: str = Field(..., min_length=2, max_length=255)
    content: str = ""
    section_order: int = 300
    section_type: HandbookSectionType = "custom"
    jurisdiction_scope: dict = Field(default_factory=dict)


class HandbookGuidedDraftRequest(BaseModel):
    title: Optional[str] = Field(default=None, max_length=500)
    mode: HandbookMode = "single_state"
    scopes: list[HandbookScopeInput] = Field(default_factory=list)
    profile: CompanyHandbookProfileInput
    industry: Optional[str] = Field(default=None, max_length=120)
    answers: dict[str, str] = Field(default_factory=dict)
    existing_custom_sections: list[HandbookGuidedSectionSuggestion] = Field(default_factory=list)


class HandbookGuidedDraftResponse(BaseModel):
    industry: str
    summary: str
    clarification_needed: bool = False
    questions: list[HandbookGuidedQuestion] = Field(default_factory=list)
    profile_updates: dict[str, Any] = Field(default_factory=dict)
    suggested_sections: list[HandbookGuidedSectionSuggestion] = Field(default_factory=list)

File: client/src/types/handbook.ts
// Handbook types — mirrors server/app/core/models/handbook.py

export type HandbookStatus = 'draft' | 'active' | 'archived'
export type HandbookMode = 'single_state' | 'multi_state'
export type HandbookSourceType = 'template' | 'upload'
export type HandbookSectionType = 'core' | 'state' | 'custom' | 'uploaded'
export type HandbookChangeStatus = 'pending' | 'accepted' | 'rejected'

export type WorkbookType =
  | 'clinical_patient_care'
  | 'infection_control'
  | 'hipaa_privacy'
  | 'safety_emergency'
  | 'human_resources'
  | 'administrative_operations'
  | 'compliance_regulatory'
  | 'financial_billing'
  | 'operations'
  | 'safety_compliance'
  | 'it_security'
  | 'finance'
  | 'general'

export const HEALTHCARE_WORKBOOK_TYPES: WorkbookType[] = [
  'clinical_patient_care',
  'infection_control',
  'hipaa_privacy',
  'safety_emergency',
  'human_resources',
  'administrative_operations',
  'compliance_regulatory',
  'financial_billing',
]

export const GENERAL_WORKBOOK_TYPES: WorkbookType[] = [
  'human_resources',
  'operations',
  'safety_compliance',
  'it_security',
  'finance',
  'general',
]

export const WORKBOOK_TYPE_LABELS: Record<WorkbookType, string> = {
  clinical_patient_care: 'Clinical / Patient Care',
  infection_control: 'Infection Control',
  hipaa_privacy: 'HIPAA & Privacy',
  safety_emergency: 'Safety & Emergency',
  human_resources: 'Human Resources',
  administrative_operations: 'Administrative / Operations',
  compliance_regulatory: 'Compliance & Regulatory',
  financial_billing: 'Financial / Billing',
  operations: 'Operations',
  safety_compliance: 'Safety & Compliance',
  it_security: 'IT & Security',
  finance: 'Finance',
  general: 'General',
}

export type HandbookScope = {
  id: string
  state: string
  city: string | null
  zipcode: string | null
  location_id: string | null
}

export type HandbookScopeInput = {
  state: string
  city?: string | null
  zipcode?: string | null
  location_id?: string | null
}

export type HandbookSection = {
  id: string
  section_key: string
  title: string
  content: string
  section_order: number
  section_type: HandbookSectionType
  jurisdiction_scope: Record<string, unknown>
  last_reviewed_at: string | null
}

export type HandbookSectionInput = {
  section_key: string
  title: string
  content: string
  section_order: number
  section_type: HandbookSectionType
  jurisdiction_scope?: Record<string, unknown> | null
}

export type CompanyHandbookProfile = {
  company_id: string
  legal_name: string
  dba: string | null
  ceo_or_president: string
  headcount: number | null
  remote_workers: boolean
  minors: boolean
  tipped_employees: boolean
  union_employees: boolean
  federal_contracts: boolean
  group_health_insurance: boolean
  background_checks: boolean
  hourly_employees: boolean
  salaried_employees: boolean
  commissioned_employees: boolean
  tip_pooling: boolean
  updated_by: string | null
  updated_at: string
}

export type CompanyHandbookProfileInput = {
  legal_name: string
  dba?: string | null
  ceo_or_president: string
  headcount?: number | null
  remote_workers: boolean
  minors: boolean
  tipped_employees: boolean
  union_employees: boolean
  federal_contracts: boolean
  group_health_insurance: boolean
  background_checks: boolean
  hourly_employees: boolean
  salaried_employees: boolean
  commissioned_employees: boolean
  tip_pooling: boolean
}

export type HandbookListItem = {
  id: string
  title: string
  status: HandbookStatus
  mode: HandbookMode
  source_type: HandbookSourceType
  active_version: number
  workbook_type: WorkbookType | null
  scope_states: string[]
  pending_changes_count: number
  updated_at: string
  published_at: string | null
  created_at: string
}

export type HandbookDetail = {
  id: string
  company_id: string
  title: string
  status: HandbookStatus
  mode: HandbookMode
  source_type: HandbookSourceType
  active_version: number
  workbook_type: WorkbookType | null
  file_url: string | null
  file_name: string | null
  scopes: HandbookScope[]
  profile: CompanyHandbookProfile
  sections: HandbookSection[]
  created_at: string
  updated_at: string
  published_at: string | null
  created_by: string | null
}

export type HandbookCreate = {
  title: string
  mode: HandbookMode
  source_type: HandbookSourceType
  industry?: string | null
  scopes: HandbookScopeInput[]
  profile: CompanyHandbookProfileInput
  custom_sections?: HandbookSectionInput[]
  guided_answers?: Record<string, string>
  file_url?: string | null
  file_name?: string | null
  create_from_template?: boolean
  auto_scope_from_employees?: boolean
  workbook_type?: WorkbookType | null
}

export type HandbookUpdate = {
  title?: string | null
  mode?: HandbookMode | null
  scopes?: HandbookScopeInput[] | null
  profile?: CompanyHandbookProfileInput | null
  sections?: HandbookSectionInput[] | null
  file_url?: string | null
  file_name?: string | null
  workbook_type?: WorkbookType | null
}

export type HandbookChangeRequest = {
  id: string
  handbook_id: string
  handbook_version_id: string
  alert_id: string | null
  section_key: string | null
  old_content: string | null
  proposed_content: string
  rationale: string | null
  source_url: string | null
  effective_date: string | null
  status: HandbookChangeStatus
  resolved_by: string | null
  resolved_at: string | null
  created_at: string
}

export type HandbookDistributionResult = {
  handbook_id: string
  handbook_version: number
  assigned_count: number
  skipped_existing_count: number
  distributed_at: string
}

export type HandbookDistributionRecipient = {
  employee_id: string
  name: string
  email: string
  invitation_status: string | null
  already_assigned: boolean
}

export type HandbookAcknowledgementSummary = {
  handbook_id: string
  handbook_version: number
  assigned_count: number
  signed_count: number
  pending_count: number
  expired_count: number
}

export type HandbookFreshnessFinding = {
  section_key: string | null
  finding_type: string
  summary: string
  change_request_id: string | null
  source_url: string | null
  effective_date: string | null
  age_days: number | null
}

export type HandbookFreshnessCheck = {
  check_id: string
  handbook_id: string
  check_type: 'manual' | 'scheduled'
  status: 'running' | 'completed' | 'failed'
  is_outdated: boolean
  impacted_sections: number
  new_change_requests_count: number
  requirements_last_updated_at: string | null
  data_staleness_days: number | null
  current_fingerprint: string | null
  previous_fingerprint: string | null
  checked_at: string
  findings: HandbookFreshnessFinding[]
}

export type HandbookCoverageByState = {
  state: string
  state_name: string
  has_addendum: boolean
  covered_categories: string[]
  missing_categories: string[]
  city_scopes: string[]
}

export type HandbookMissingSection = {
  section_key: string
  title: string
  reason: string
  priority: string
}

export type HandbookCoverage = {
  handbook_id: string
  strength_score: number
  strength_label: string
  total_sections: number
  core_sections: number
  state_sections: number
  custom_sections: number
  uploaded_sections: number
  federal_core_count: number
  state_level_count: number
  city_level_count: number
  state_coverage: HandbookCoverageByState[]
  missing_sections: HandbookMissingSection[]
  industry: string
  industry_label: string
}

export type HandbookGuidedQuestion = {
  id: string
  question: string
  placeholder: string | null
  required: boolean
}

export type HandbookGuidedSectionSuggestion = {
  section_key: string
  title: string
  content: string
  section_order: number
  section_type: HandbookSectionType
  jurisdiction_scope: Record<string, unknown>
}

export type HandbookGuidedDraftRequest = {
  title?: string | null
  mode: HandbookMode
  scopes: HandbookScopeInput[]
  profile: CompanyHandbookProfileInput
  industry?: string | null
  answers: Record<string, string>
  existing_custom_sections?: HandbookGuidedSectionSuggestion[]
}

export type HandbookGuidedDraftResponse = {
  industry: string
  summary: string
  clarification_needed: boolean
  questions: HandbookGuidedQuestion[]
  profile_updates: Record<string, unknown>
  suggested_sections: HandbookGuidedSectionSuggestion[]
}

export type HandbookWizardDraftState = Record<string, unknown>

export type HandbookWizardDraft = {
  id: string
  company_id: string
  user_id: string
  state: HandbookWizardDraftState
  created_at: string
  updated_at: string
}

export type HandbookPublishResponse = {
  id: string
  status: HandbookStatus
  active_version: number
  published_at: string | null
}
